src/compute-client/src/as_of_selection.rs

```diff
@@ -0,0 +1,521 @@
+    ) {
+        // Apply direct constraints from storage inputs.
+        for (id, collection) in &self.collections {
+            for input_id in &collection.storage_inputs {
+                let read_hold = &storage_read_holds[input_id];
+                let constraint = Constraint {
+                    type_: ConstraintType::Hard,
+                    bound_type: BoundType::Lower,
+                    frontier: read_hold.since(),
+                    reason: &format!("storage input {input_id} read frontier"),
+                };
+                self.apply_constraint(*id, constraint);
+            }
+        }
+
+        // Propagate constraints downstream, restoring `AsOfBounds` invariant (1).
+        self.propagate_bounds_downstream(BoundType::Lower, ConstraintType::Hard);
+    }
+
+    /// Apply as-of constraints imposed by the frontiers of downstream storage collections.
+    ///
+    /// A collection's as-of _must_ be <= the join of the read frontier and the write frontier of
+    /// the storage collection it exports to, if any.
+    ///
+    /// Failing to apply this constraint to a collection is an error. The storage collection it
+    /// exports to will have times visible to readers skipped in its output, violating correctness.
+    fn apply_downstream_storage_constraints(&self) {
+        // Apply direct constraints from storage exports.
+        for id in self.collections.keys() {
+            let Ok(frontiers) = self.storage_collections.collection_frontiers(*id) else {
+                continue;
+            };
+            let upper = frontiers.read_capabilities.join(&frontiers.write_frontier);
+            let constraint = Constraint {
+                type_: ConstraintType::Hard,
+                bound_type: BoundType::Upper,
+                frontier: &upper,
+                reason: &format!("storage export {id} write frontier"),
+            };
+            self.apply_constraint(*id, constraint);
+        }
+
+        // Propagate constraints upstream, restoring `AsOfBounds` invariant (2).
+        self.propagate_bounds_upstream(BoundType::Upper, ConstraintType::Hard);
+    }
+
+    /// Propagate as-of bounds through the dependency graph, in downstream direction.
+    ///
+    /// Calling this method restores `AsOfBounds` invariant (2).
+    fn propagate_bounds_downstream(&self, bound_type: BoundType, constraint_type: ConstraintType) {
+        // We don't want to rely on a correspondence between `GlobalId` order and dependency order,
+        // so we use a fixpoint loop here.
+        fixpoint(|changed| {
+            self.propagate_bounds_downstream_inner(bound_type, constraint_type, changed);
+
+            // Propagating `upper` bounds downstream might break `AsOfBounds` invariant (2), so we
+            // need to restore it.
+            if bound_type == BoundType::Upper {
+                self.propagate_bounds_upstream_inner(
+                    BoundType::Upper,
+                    ConstraintType::Hard,
```

Comment:
This is supposed to restore one of the invariants. If it fails, that means one of the invariants is violated and we will probably end up with dataflows that can't hydrate or skip times in their output, which would be a correctness bug.

Reply:
Ah yes, ok. (When I wrote this comment, I mistakenly thought that `AsOfBounds` tracks soft and hard bounds separately, and then propagating soft constrains would have affected only soft bounds, so invariants would have had to be restored only for the soft bounds. But actually, this is not how this works; the soft-hard thing is only for constraints.)

Reply:
Good question! I've convinced myself that this can't happen, but let me try to convince you too.

Let's assume we have just two collections, `A --> B`

* When we start, both invariants are upheld, i.e. `A.lower <= B.lower` and `A.upper <= B.upper`.
* Case 1: soft constraint increases `A.lower`
  * We know that `A.lower` can only be increased up to `A.upper`, so `A.lower' <= A.upper <= B.upper`.
  * Since `A.lower' <= B.upper`, we can propagate the new lower bound to `B`, restoring the invariant.
* Case 2: soft constraint decreases `A.upper`
  * `A.upper' < A.upper <= B.upper`, so the invariant is still upheld without any propagation.
* Case 3: soft constraint increases `B.lower`
  * `A.lower <= B.lower < B.lower'`, so the invariant is still upheld without any propagation.
* Case 4: soft constraint decreases `B.upper`
  * We know that `B.upper` can only be decreased up to `B.lower`, so `A.lower <= B.lower <= B.upper'`.
  * Since `A.lower <= B.upper'`, we can propagate the new upper bound to `A`, restoring the invariant.

Unless I missed something, this means that applying a soft constraint is always safe, as long as one only applies a single constraint at a time and then propagates. We are not doing that, we are always applying one constraint to all collections and then propagate. I think this is still fine as long as a batch of constraint application targets the same bound type (lower or upper).

* Case 5: soft constraint increases both `A.lower` and `B.lower`
  * Case 5.a: `A.lower' <= B.lower'`
    *  The invariant is still upheld without propagation.
  * Case 5.b: `A.lower' > B.lower'`
    * Same as case 1.
* Case 6: soft constraint decreases both `A.upper` and `B.upper`
  * Case 6.a: `A.upper' <= B.upper'`
    *  The invariant is still upheld without propagation.
  * Case 6.b: `A.upper' > B.upper'`
    * Same as case 4.

Reply:
Ok, I'm convinced! (But the first sentence of the proof seems a bit misleading: I think we don't really want to assume that we have just two collections. Instead, we are picking an arbitrary dependency edge `A --> B`, and proving that after a constraint changes the bounds on `A` or `B`, propagation will always succeed.)

Also, the proof seems to apply not just to these extra `ConstraintType::Hard` calls in `propagate_bounds_downstream` and `propagate_bounds_upstream`, which were the subject of my original comment, but also apply to all `apply_constraint` calls that happen in `propagate_bounds_downstream_inner` and `propagate_bounds_upstream_inner`, right? In other words, no constraint application can ever fail inside any of the `propagate_*` functions. If this is true, then could we remove the `constraint_type` argument of the `propagate_*` functions and make all `apply_constraint` calls in `propagate_*` functions with a `Hard` constraint? 

(Or maybe even create an even harder constraint type, whose failing application would signal an internal error specifically in this module. This would be a more specific error than a normal hard constraint application fail, because a normal hard constraint application fail might also result from, say, corruption in the persisted collection state, if I'm not mistaken.)

Reply:
> (But the first sentence of the proof seems a bit misleading: I think we don't really want to assume that we have just two collections. Instead, we are picking an arbitrary dependency edge `A --> B`, and proving that after a constraint changes the bounds on `A` or `B`, propagation will always succeed.)

Yeah, sorry. I meant to say that this holds for arbitrary dependent collections, and thereby also holds for arbitrary dependency graphs.

> If this is true, then could we remove the `constraint_type` argument of the `propagate_*` functions and make all `apply_constraint` calls in `propagate_*` functions with a `Hard` constraint?

Oh, nice catch! Yeah, I think we could. I'll just make them `Hard` constraints instead of introducing a new constraint type, since the `reason` should already give us specific errors.

Reply:
And could you add the proof as a code comment?

Reply:
Made the `propagate_bounds_*` methods always use hard constraints and added the proof to the module docstring.

Reply:
> If this is true, then could we remove the constraint_type argument of the propagate_* functions and make all apply_constraint calls in propagate_* functions with a Hard constraint?

Turns out I we can't just remove the `constraint_type` from the `propagate_bounds_*` methods. It's true that restoring the invariants should always succeed, but `propagate_bounds_*` lets you also do propagations in the other direction from what restores the invariant. For example, the step that applies warmup constraints applies `Upper` constraints to the collections reading from storage directly, and then propagates the `Upper` bounds _downstream_, i.e. it tries to hold back the as-of of downstream collections to the as-ofs of their inputs. Doing so is not important for the invariants, so it's fine if that fails.

I think we can derive the `constraint_type` from the direction and the `bounds_type` though, based on whether they restore an invariant or not. I.e.:
* downstream, `Lower` => `Hard`
* downstream, `Upper` => `Soft`
* upstream, `Lower` => `Soft`
* upstream, `Upper` => `Hard`